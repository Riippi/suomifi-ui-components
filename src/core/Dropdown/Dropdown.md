```js
import { Dropdown } from 'suomifi-ui-components';

<Dropdown
  name="Dropdown"
  labelText="Dropdown label"
  defaultValue={'Dropdown item 1'}
  onChange={(newValue) => console.log(newValue)}
>
  <Dropdown.item value={'Dropdown item 1'}>
    Dropdown Item 1
  </Dropdown.item>
  <Dropdown.item value={'Dropdown item 2'}>
    Dropdown Item 2
  </Dropdown.item>
</Dropdown>;
```

```js
import { Dropdown } from 'suomifi-ui-components';

<Dropdown
  name="Dropdown"
  visualPlaceholder="Dropdown with visually hidden label"
  labelText="Dropdown label"
  labelMode="hidden"
>
  <Dropdown.item value={'Dropdown item 1'}>
    Dropdown Item 1
  </Dropdown.item>
  <Dropdown.item value={'Dropdown item 2'}>
    Dropdown Item 2
  </Dropdown.item>
</Dropdown>;
```

```js
import { Dropdown } from 'suomifi-ui-components';

<Dropdown
  visualPlaceholder="Action menu"
  labelText="Dropdown as action menu label"
  alwaysShowVisualPlaceholder={true}
  onChange={(action) => console.log(action, 'selected')}
>
  <Dropdown.item value={'Action item 1'}>Action Item 1</Dropdown.item>
  <Dropdown.item value={'Action item 2'}>Action Item 2</Dropdown.item>
</Dropdown>;
```

```js
import { useState } from 'react';
import { Dropdown } from 'suomifi-ui-components';

const [value, setValue] = useState(undefined);

<Dropdown
  value={value}
  visualPlaceholder="Dropdown"
  labelText="Dropdown with controlled state"
  onChange={(newValue) => {
    if (window.confirm('Change dropdown value?')) {
      setValue(newValue);
    }
  }}
>
  <Dropdown.item value={'Action item 1'}>Action Item 1</Dropdown.item>
  <Dropdown.item value={'Action item 2'}>Action Item 2</Dropdown.item>
</Dropdown>;
```

```js
import { Dropdown, Block } from 'suomifi-ui-components';

const dropdownProps = {
  'aria-labelledby': 'dropdown-group',
  onChange: (newValue) => console.log(newValue)
};

<div>
  <label id="dropdown-group">
    <p>Dropdown group</p>
  </label>
  <Block padding="xs">
    <Dropdown
      defaultValue={'Dropdown 1 item 2'}
      visualPlaceholder="Dropdown 1"
      labelText="Dropdown 1 label"
      {...dropdownProps}
    >
      <Dropdown.item value={'Dropdown 1 item 1'}>
        Dropdown 1 Item 1
      </Dropdown.item>
      <Dropdown.item value={'Dropdown 1 item 2'}>
        Dropdown 1 Item 2
      </Dropdown.item>
    </Dropdown>
  </Block>
  <Block padding="xs">
    <Dropdown
      defaultValue={'Dropdown 2 item 2'}
      visualPlaceholder="Dropdown 2"
      labelText="Dropdown 2 label"
      {...dropdownProps}
    >
      <Dropdown.item value={'Dropdown 2 item 1'}>
        Dropdown 2 Item 1
      </Dropdown.item>
      <Dropdown.item value={'Dropdown 2 item 2'}>
        Dropdown 2 Item 2
      </Dropdown.item>
    </Dropdown>
  </Block>
</div>;
```
