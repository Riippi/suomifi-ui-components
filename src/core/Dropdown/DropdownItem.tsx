import React, { ReactNode } from 'react';
import { ListboxOption } from '@reach/listbox';

export interface DropdownItemProps {
  /** Item value */
  value: string;
  /** Item content */
  children: ReactNode;
  /** Classname for item */
  className?: string;
}

export const DropdownItem = (props: DropdownItemProps) => (
  <ListboxOption {...props} />
);
