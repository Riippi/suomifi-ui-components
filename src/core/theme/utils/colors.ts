import { ColorDesignTokens } from 'suomifi-design-tokens';
import { InternalTokensProp } from '../';

export const colorValue = (props: InternalTokensProp) => (
  colorToken: keyof ColorDesignTokens,
) => props.tokens.colors[colorToken].hsl;

export const colorsUtils = (colors: ColorDesignTokens) =>
  Object.entries(colors).reduce(
    (retObj, [key, value]) => ({
      ...retObj,
      [key]: value.hsl,
    }),
    {} as { [key in keyof ColorDesignTokens]: string },
  );
