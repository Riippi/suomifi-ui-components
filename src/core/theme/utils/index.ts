export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { spacingUtils, margin, padding, spacingModifiers } from './spacing';
export {
  typographyUtils,
  TypographyUtil,
  fontSize,
  FontSizeProps,
} from './typography';
export { colorsUtils, colorValue } from './colors';
