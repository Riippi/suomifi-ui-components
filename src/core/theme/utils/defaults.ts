import {
  suomifiTheme,
  SuomifiTheme,
  defaultTokens,
  SuomifiTokens,
  InternalTokensProp,
  PartialTokens,
} from '../';
import { asPropType, Omit } from '../../../utils/typescript';

const internalTokens = (tokens?: PartialTokens) =>
  !!tokens ? { ...defaultTokens, ...tokens } : defaultTokens;

/**
 * Check component props and do common defaulting
 * - Check if tokens are given or use default ones
 * - include internal tokens (THESE ARE NOT GEMERATED BY GIVEN TOKENS! but can be overridden by given tokens)
 * - If as-prop (styled-components) set pass it as asProp
 * @param props All component's props
 */
export const withSuomifiDefaultProps = <
  T extends { tokens?: PartialTokens; asProp?: asPropType }
>({
  tokens,
  as,
  ...props
}: T & { as?: asPropType }) =>
  ({
    ...props,
    tokens: internalTokens(tokens),
    ...(!!as ? { asProp: as } : {}),
  } as Omit<T, 'tokens'> & InternalTokensProp);

export interface Tokens {
  tokens: SuomifiTokens;
}

export interface Theme {
  theme: SuomifiTheme;
}
export type TokensOrThemeProps = Tokens | Theme;

/**
 * Return theme if defined or theme based on tokens
 * @param {Object} prop.theme Theme if defined
 * @param {Object} prop.tokens Tokens
 */
export const themeOrTokens = (props: TokensOrThemeProps) =>
  'theme' in props ? props.theme : suomifiTheme(props.tokens);
