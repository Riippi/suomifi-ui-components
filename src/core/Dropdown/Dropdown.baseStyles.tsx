import { css } from 'styled-components';
import { suomifiTheme, ThemeProp } from '../theme';
import { DropdownProps } from './Dropdown';
import { element, inputButton, fonts } from '../theme/reset';
import { Omit } from '../../utils/typescript';

export const baseStyles = ({
  theme = suomifiTheme,
}: Omit<DropdownProps, 'name'>) => css`
  & > [data-reach-menu-button].fi-dropdown_button {
    ${inputButton(theme)}
    position: relative;
    padding-right: 30px;
    text-align: left;
    cursor: pointer;
    &:before {
      content: '';
      position: absolute;
      top: 50%;
      right: 10px;
      margin-top: -3px;
      border-style: solid;
      border-color: ${theme.colors.depthDark27} transparent transparent
        transparent;
      border-width: 6px 4px 0 4px;
    }
    &[aria-expanded='true']:before {
      border-color: transparent transparent ${theme.colors.depthDark27}
        transparent;
      border-width: 0 4px 6px 4px;
    }
  }
`;

export const menuListStyles = ({
  theme = suomifiTheme,
}: {
  theme: ThemeProp;
}) => css`
  &[data-reach-menu-list].fi-dropdown_list {
    ${element(theme)}
    ${fonts(theme).input}
    margin-top: -1px;
    padding: 0;
    font-size: 100%;
    border: 0;
    background-color: ${theme.colors.whiteBase};
    border-color: ${theme.colors.depthBase};
    border-style: solid;
    border-width: 0 1px 1px 1px;
    border-radius: 0px 0px ${theme.radius.basic} ${theme.radius.basic};
    overflow: hidden;
  }

  & [data-reach-menu-item].fi-dropdown_item {
    ${element(theme)}
    ${fonts(theme).input}
    padding: ${theme.spacing.s} ${theme.spacing.m};
    border: 0;
    &[data-selected] {
      ${fonts(theme).input}
      color: ${theme.colors.blackBase};
      background-image: none;
      background-color: ${theme.colors.highlightLight50};
      border: 0;
    }
  }
`;
