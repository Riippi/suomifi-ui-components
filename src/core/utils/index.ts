import { FlattenSimpleInterpolation } from 'styled-components';
import { suomifiTheme, ThemeProp } from '../theme';

interface BaseStylesInterface {
  theme: ThemeProp;
  [key: string]: any;
}

/**
 *  Get CSS export from baseStyles generated by Emotion/Styled
 * @param baseStyles Emotion CSS
 * @param props properties needed to get desired styles from components
 */
export const cssFromBaseStyles = (
  baseStyles: ({ theme }: BaseStylesInterface) => FlattenSimpleInterpolation,
  props = {},
) => baseStyles({ ...props, theme: suomifiTheme }).join('');
