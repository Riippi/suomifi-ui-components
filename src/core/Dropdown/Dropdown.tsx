import React, { Component, Fragment } from 'react';
import { default as styled } from 'styled-components';
import { withSuomifiDefaultProps } from '../theme/utils';
import { TokensProp, InternalTokensProp } from '../theme';
import { baseStyles, menuPopoverStyles } from './Dropdown.baseStyles';
import {
  MenuPopover as CompMenuPopover,
  MenuPopoverProps as CompMenuPopoverProps,
  MenuItems as CompMenuItems,
} from '../../components/LanguageMenu/LanguageMenu';
import {
  Dropdown as CompDropdown,
  DropdownProps as CompDropdownProps,
} from '../../components/Dropdown/Dropdown';
import { DropdownItem, DropdownItemProps } from './DropdownItem';
import { positionMatchWidth } from '@reach/popover';

export interface DropdownProps extends CompDropdownProps, TokensProp {}

const StyledDropdown = styled(
  ({ tokens, ...passProps }: DropdownProps & InternalTokensProp) => (
    <CompDropdown
      {...passProps}
      dropdownItemProps={{ className: 'fi-dropdown_item' }}
    />
  ),
)`
  ${props => baseStyles(props)}
`;

interface MenuPopoverProps extends CompMenuPopoverProps, TokensProp {}

const StyledMenuPopover = styled(
  ({ tokens, children, ...passProps }: MenuPopoverProps) => (
    <CompMenuPopover position={positionMatchWidth} {...passProps}>
      <CompMenuItems>{children}</CompMenuItems>
    </CompMenuPopover>
  ),
)`
  ${props => menuPopoverStyles(props.theme)}
`;

/**
 * <i class="semantics" />
 * Use for selectable dropdown with items.
 */
export class Dropdown extends Component<DropdownProps> {
  static item = (props: DropdownItemProps) => <DropdownItem {...props} />;

  render() {
    const props = withSuomifiDefaultProps(this.props);
    return (
      <Fragment>
        <StyledDropdown {...props} menuPopoverComponent={StyledMenuPopover} />
      </Fragment>
    );
  }
}
