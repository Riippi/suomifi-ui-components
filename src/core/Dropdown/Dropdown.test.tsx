import React from 'react';
import { cssFromBaseStyles } from '../utils';
import { baseStyles } from './Dropdown.baseStyles';
import { axeTest } from '../../utils/test/axe';

import { Dropdown } from './Dropdown';

const doNothing = () => ({});

const TestDropdown = (
  <Dropdown className="dropdown-test" name="Dropdown">
    <Dropdown.item onSelect={doNothing}>Item 1</Dropdown.item>
    <Dropdown.item onSelect={doNothing}>Item 2</Dropdown.item>
  </Dropdown>
);

test('CSS export', () => {
  const css = cssFromBaseStyles(baseStyles);
  expect(css).toEqual(expect.stringContaining('position'));
});

test('should not have basic accessibility issues', axeTest(TestDropdown));
