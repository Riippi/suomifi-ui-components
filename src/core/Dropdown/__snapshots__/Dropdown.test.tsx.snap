// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c1 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0 > [data-reach-menu-button].fi-dropdown_button {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  min-width: 245px;
  max-width: 100%;
  padding: 8px 16px;
  border: 1px solid hsl(202,7%,67%);
  border-radius: 2px;
  line-height: 1;
  position: relative;
  padding-right: 30px;
  text-align: left;
  background-color: hsl(0,0%,100%);
  cursor: pointer;
}

.c0 > [data-reach-menu-button].fi-dropdown_button:focus {
  outline: 0;
  position: relative;
}

.c0 > [data-reach-menu-button].fi-dropdown_button:focus:after {
  content: '';
  position: absolute;
  top: -4px;
  right: -4px;
  bottom: -4px;
  left: -4px;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid hsl(23,82%,53%);
  box-sizing: border-box;
  box-shadow: 0 0 10px 0 hsl(23,82%,53%);
  z-index: 9999;
}

.c0 > [data-reach-menu-button].fi-dropdown_button:focus:not(:focus-visible):after {
  content: none;
}

.c0 > [data-reach-menu-button].fi-dropdown_button:before {
  content: '';
  position: absolute;
  top: 50%;
  right: 10px;
  margin-top: -3px;
  border-style: solid;
  border-color: hsl(202,7%,40%) transparent transparent transparent;
  border-width: 6px 4px 0 4px;
}

.c0 > [data-reach-menu-button].fi-dropdown_button[aria-expanded='true']:before {
  border-color: transparent transparent hsl(202,7%,40%) transparent;
  border-width: 0 4px 6px 4px;
}

<div>
  <span
    class="dropdown-test c0 fi-dropdown c1"
  >
    <button
      aria-controls="menu--1"
      aria-haspopup="true"
      class="fi-dropdown_button"
      data-reach-menu-button=""
      id="menu-button--menu--1"
      type="button"
    >
      Dropdown
    </button>
  </span>
</div>
`;
